lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -59,6 +62,98 @@ static StringAttr filterUselessName(StringAttr name) {
   return isUselessName(name.getValue()) ? StringAttr() : name;
 }
 
+enum AnnotationID {
+  InlineAnnotationID,
+  NoDedupAnnotationID,
+  RunFirrtlTransformAnnotationID,
+  // An annotation that we know about, but haven't ported yet
+  UnsupportedAnnotationID,
+  // An annotation that we don't know what to do with
+  UnknownAnnotationID
+};
+
+LogicalResult
+fromJSON(llvm::json::Value &value,
+         llvm::StringMap<llvm::SmallVector<Annotation>> &annotationMap,
+         llvm::json::Path path, MLIRContext *context) {
+
+  auto findTarget =
+      [](llvm::json::Object *object) -> llvm::Optional<std::string> {
+    auto maybeTarget = object->get("target")->getAsString();
+
+    // This is a normal target. Just return it.
+    if (maybeTarget.getValue()[0] == '~')
+      return llvm::Optional<std::string>(maybeTarget->str());
+
+    // This is a legacy "named" target and is canonicalized to a target
+    // here. This converts:
+    //   Foo.Bar.Baz -> ~Foo|Bar>Baz
+    std::string newTarget = "~";
+    llvm::raw_string_ostream s(newTarget);
+    bool isModule = true;
+    for (auto a : maybeTarget.getValue()) {
+      switch (a) {
+      case '.':
+        if (isModule) {
+          s << "|";
+          isModule = false;
+          break;
+        }
+        s << ">";
+        break;
+      default:
+        s << a;
+      }
+    }
+    return llvm::Optional<std::string>(newTarget);
+  };
+
+  if (auto object = value.getAsObject()) {
+    if (auto clazz = object->get("class")) {
+      if (auto className = clazz->getAsString()) {
+        auto id =
+            llvm::StringSwitch<AnnotationID>(className.getValue())
+                .Case("firrtl.passes.InlineAnnotation", InlineAnnotationID)
+                .Case("firrtl.transforms.NoDedupAnnotation",
+                      NoDedupAnnotationID)
+                .Case("firrtl.stage.RunFirrtlTransformAnnotation",
+                      RunFirrtlTransformAnnotationID)
+                .Default(UnknownAnnotationID);
+        switch (id) {
+        case InlineAnnotationID:
+          annotationMap[findTarget(object).getValue()].push_back(
+              InlineAnnotation::get(context));
+          break;
+        case NoDedupAnnotationID:
+          annotationMap[findTarget(object).getValue()].push_back(
+              NoDedupAnnotation::get(context));
+          break;
+        case RunFirrtlTransformAnnotationID:
+          annotationMap["~"].push_back(RunFirrtlTransformAnnotation::get(
+              context,
+              object->get("transform")->getAsString().getValue().str()));
+          break;
+        case UnsupportedAnnotationID:
+          llvm::errs() << "Unsupported annotation: " << className << "\n";
+          return failure();
+          break;
+        case UnknownAnnotationID:
+          llvm::errs() << "Unknown annotation: " << className << "\n";
```

Comment:
The JSON parser doesn't support tracking file locations at all ;-( but it is untidy to emit errors with `errs`.  I'd recommend passing a location of the circuit declaration in the .fir file or something, and then emitError'ing on it.  This will allow us to filecheck these things etc.

Reply:
I can switch to that. I'll probably just parse the circuit definition, parse the annotation file (erroring pointing at the circuit location) to build up the annotation data structure, then proceed to parse everything.

Reply:
Fixed in newer squash commits. The PR now delays processing the input annotation file until it has a circuit. The error is now associated with the circuit op. A JSON parser error failure isn't amazing: